src/main/kotlin/com/wgslfuzz/core/Resolver.kt

```diff
@@ -765,15 +821,20 @@ private fun resolveUnary(
         if (pointerType !is Type.Pointer) {
             throw RuntimeException("Dereference applied to expression $expression with non-pointer type")
         }
-        pointerType.pointeeType
+        Type.Reference(pointerType.pointeeType, pointerType.addressSpace, pointerType.accessMode)
     }
 
     UnaryOperator.ADDRESS_OF -> {
-        resolveTypeOfAddressOfExpression(expression, resolverState)
+        val referenceType = resolverState.resolvedEnvironment.typeOf(expression.target)
+        if (referenceType !is Type.Reference) {
+            throw RuntimeException("Address-of applied to expression $expression with non-reference type")
+        }
+        Type.Pointer(referenceType.storeType, referenceType.addressSpace, referenceType.accessMode)
     }
 
     UnaryOperator.LOGICAL_NOT -> {
-        assert(resolverState.resolvedEnvironment.typeOf(expression.target) == Type.Bool)
+        val targetType = resolverState.resolvedEnvironment.typeOf(expression.target)
+        assert(targetType == Type.Bool || (targetType is Type.Reference && targetType.storeType is Type.Bool))
```

Comment:
Could you change this to a check that will throw an exception, rather than an assertion?

I have not been very consistent about whether to assert things, or throw an exception, and if the latter which kind of exception, so any cleanup on that front you can do would be great.